tokio-tasks/src/task_map.rs

```diff
@@ -0,0 +1,71 @@
+use crate::future_ext::FutureExt;
+use futures::future::RemoteHandle;
+use futures::Future;
+use std::cmp;
+use std::collections::HashMap;
+use std::hash;
+
+/// Struct controlling the lifetime of the async tasks, such as
+/// running actors and periodic notifications. If it gets dropped, all
+/// tasks are cancelled.
+///
+/// Each task is associated with a key, so that it can be removed from
+/// the internal `HashMap` whenever consumers want to stop tracking
+/// the corresponding task. By removing it from the `HashMap` we free
+/// up space.
+pub struct TaskMap<K>(HashMap<K, RemoteHandle<()>>);
```

Comment:
Could we use this one directly instead?

Reply:
I think it would be dramatically less scalable because we would have to block the processing of `Ping` messages based on `Finished`.

At this stage I see actors that handle messages that access different data and are completely independent as an antipattern.

Reply:
Oh, this *was* the added value! I missed the crucial bit! Indeed, the `Spawner` makes sense - but then, do we really need to have the `TaskMap` and the keyed usage?

Reply:
The way I saw it, we need to send ourselves a message informing of a task having finished and we need to be able to know which task handle to drop based on that.

What do you propose other than something like `TaskMap`?

Reply:
hm. coming at it again, the fact that `Spawner` uses a TaskMap is its implementation detail. We could generate unique id of a future when it's started, instead of passing it in via the message from the outside. This way it would be easier to use and also misuse-resistant (no arbitrary replacement of the same keys, dropping a handle mid-way through.

Reply:
That's a good idea, I like that!